File: May 2022/May_13.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if(root == nullptr)
            return root;
        queue<Node* > q;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            Node* prev = nullptr;
            while(size--){
                Node* rem = q.front();
                q.pop();
                if(prev != nullptr)
                    prev->next = rem;
                prev = rem;
                if(rem->left != nullptr)
                    q.push(rem->left);
                if(rem->right != nullptr)
                    q.push(rem->right);
            }
            prev->next = nullptr;
        }
        return root;
    }
};

File: May 2022/May_13.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null)
            return root;
        Queue<Node> q = new ArrayDeque<>();
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            Node prev = null;
            while(size-- > 0){
                Node rem = q.remove();
                if(prev != null)
                    prev.next = rem;
                prev = rem;
                if(rem.left != null)
                    q.add(rem.left);
                if(rem.right != null)
                    q.add(rem.right);
            }
            prev.next = null;
        }
        return root;
    }
}
